### PR TITLE
KAFKA-3428 Remove metadata sync bottleneck from mirrormaker's producer

### PR DESCRIPTION
**Pull request by maysamyabandeh**
Repalce topics with a concurrent hashset so it would not require to be
modified inside a synchrnoized method.
Make cluster a volatile varialble so fetch which just returns the
pointer does not have to be synchrnized.

@ijuma @becketqin

the contribution is my original work and that i license the work to the project under the project's open source license.

### CODE REVIEW

**Comment by maysamyabandeh** [comment]
@ijuma would you be interested in reviewing this pull request?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 1]
```diff
@@ -166,7 +167,10 @@ public synchronized void update(Cluster cluster, long now) {
             listener.onMetadataUpdate(cluster);
 
         // Do this after notifying listeners as subscribed topics' list can be changed by listeners
-        this.cluster = this.needMetadataForAllTopics ? getClusterForCurrentTopics(cluster) : cluster;
+        final Set<String> topicsCopy = topics();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Is this really needed? Aren't all modifications of `topics` done within a `synchronized` block?

**Comment by ijuma** [comment]
Thanks for the PR. I think we need to document in the code the reasoning for the proposed strategy where most fields are protected by synchronization but a couple are not.

It would also be interesting to know if the consumer also suffers from similar bottlenecks and, if so, whether these change fix that case too.

What are your thoughts @guozhangwang @ewencp? For reference, I quote from the JIRA the relevant figures:

"Due to sync on the single producer, MM in a setup with 32 consumer threads could not send more than 
358k msg/sec hence not being able to saturate the NIC."

"After applying the patch, MM could copy the entire data (447k msg/sec) with no lagging. The bytes-in rate was 183m per sec. We have not tried it yet with an input rate large enough to saturate a 10G NIC and we might discover other bottlenecks when doing so."

**Comment by ijuma** [comment]
Also, it would be good to verify that changing `topics` to a Concurrent implementation doesn't slow down the case where there are less threads.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 1 above)
Comment (maysamyabandeh) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: good point. it was overly cautious. will remove that.

**Comment by maysamyabandeh** [comment]
fyi, with this change we got around 1M msg/s (varies on different setup), which is still not high enough to saturate the NIC but still a non-negligible improvement.

we could not identify an obvious bottleneck after this point. the time consuming operations seems to be synchronization on free and incomplete, as well as synchronization on partition dequeue.

**Comment by ewencp** [comment]
I think its great to see performance optimizations like this, but agreed with @ijuma that this definitely needs documentation about the approach and reasoning (and probably a reference to the relevant JIRA). The synchronized approach is clearly simpler to reason about and without any notes in the code it'd be really easy to revert important changes here. (Ideally we'd also have tests that could detect performance regression here. We don't have real support for this in the system tests yet, but this is a pretty specific scenario we'd need to test -- very large # of threads utilizing a single producer.)

**Comment by maysamyabandeh** [comment]
Sure. I am about to submit a new PR with documentation.

About concerns of ConcurrentHashMap's read performance, I did a quick test with a single consumer and both approaches gave about 194 Msg/s. Although the experiment was performed only once and number varied over time, it suggests that ConcurrentHashMap's read latency is negligible compared to the overall latency of invoking producer.send

Still if that is a concern we have the option of making topics volatile and replace the reference with a new updated HashMap but i don't see evidence that this would be necessary.

**Comment by maysamyabandeh** [comment]
not sure why is Jenkins [complaining](https://builds.apache.org/job/kafka-trunk-git-pr-jdk7/3128/testReport/kafka.api/SaslPlaintextConsumerTest/testSimpleConsumption/) about this test:

> kafka.api.SaslPlaintextConsumerTest > testSimpleConsumption FAILED
>     java.lang.AssertionError: expected:<1> but was:<0>

I ran the test on my laptop and passed successfully!

**Comment by maysamyabandeh** [comment]
@ijuma @ewencp are you guys comfortable with the latest patch?

**Comment by maysamyabandeh** [comment]
@ijuma @ewencp a kindly reminder

File: gradlew [hunk 2]
```diff
@@ -42,6 +42,11 @@ case "`uname`" in
     ;;
 esac
 
+# For Cygwin, ensure paths are in UNIX format before anything is touched.
```
Comment (ijuma) [inline gradlew]: This change is unrelated.

File: gradlew.bat [hunk 3]
```diff
@@ -46,7 +46,7 @@ echo location of your Java installation.
 goto fail
 
 :init
-@rem Get command-line arguments, handling Windows variants
+@rem Get command-line arguments, handling Windowz variants
```
Comment (ijuma) [inline gradlew.bat]: This change is unrelated.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 4]
```diff
@@ -42,8 +43,17 @@
...
+     * KAFKA-3428: Cluster class is assumed to be safe for concurrent access. To allow high concurrency on fetch(), 
+     * which returns the pointer to cluster, this method is not synchronized and safety is assured by cluster being a volatile pointer.
+     */
+    private volatile Cluster cluster;
     private boolean needUpdate;
+    /**
+     * KAFKA-3428
+     * synchronization policy: topics is updated only within Metadata synchronized methods. The read queries to topics however
+     * are not protected by external locks and topics must use an implementation that is safe to do reads concurrent with updates.
+     */
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Instead of spreading these comments in many places, I would have a single comment on the line after the opening class brace explaining the general strategy. Also, I would move the two fields that are accessed without a lock into a separate group with a brief comment just before them.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 5]
```diff
@@ -69,15 +79,18 @@ public Metadata(long refreshBackoffMs, long metadataExpireMs) {
         this.version = 0;
         this.cluster = Cluster.empty();
         this.needUpdate = false;
-        this.topics = new HashSet<String>();
+        // KAFKA-3428: using concurrent set since reads to topics can be concurrent with writes.
+        this.topics = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
         this.listeners = new ArrayList<>();
         this.needMetadataForAllTopics = false;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: I would move all of these apart from the first two to the declaration of the fields themselves.

**Comment by vvcephei** [comment]
Hi @maysamyabandeh ,

It seems like this PR stalled. I'll close it out for now, but if you or anyone else want to resume this work, please feel free to re-open it (or start a new one)!

Thanks,
John
